File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/22-PQ/LeetCode/LeetCode-2530.cpp
class Solution {
public:
    /// O(n+klogn) with O(n) extra space
    long long maxKelements(vector<int>& nums, int k) {        
        priority_queue<int> Q(nums.begin(), nums.end());

        int64_t result = 0;
        while (k--){
            int num = Q.top();
            Q.pop();
            result += num;
            Q.push((num+2)/3);
        } // end-while
        return result;
    } // end-maxKelements
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/22-PQ/LeetCode/LeetCode-2530.java
class Solution {
    /// O(nlogn + klogn) algorithm with O(n) extra space
    public long maxKelements(int[] nums, int k) {
        /// Create a max PQ and insert all numbers
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>((x, y)->Integer.compare(y, x));
        for (var num: nums) pq.offer(num);

        long result = 0;
        while (k-- > 0){
            var val = pq.poll();
            result += val;
            pq.offer((val+2)/3);
        } // end-while
        return result;        
    }
}
